apispec-model/src/main/scala/sttp/apispec/validation/SchemaComparator.scala

```diff
@@ -0,0 +1,534 @@
+   * the comparator simply falls back to comparing schemas by plain equality, or returns a [[GeneralSchemaMismatch]],
+   * which indicates comparator's inability to definitely determine compatibility or incompatibility of the schemas.
+   *
+   * In practice, the comparator is designed to detect typical changes that may appear in schemas during API evolution,
+   * e.g. adding new fields, changing types, etc.
+   *
+   * Before being compared, all schemas are stripped of keywords which do not affect the comparison, e.g. annotations
+   * like `title`, `description`, etc.
+   *
+   * @param writerSchema       schema of the data being written
+   * @param readerSchema       schema of the data being read
+   * @param writerNamedSchemas named schemas which may be referred to by the writer schema
+   * @param readerNamedSchemas named schemas which may be referred to by the reader schema
+   * @return a list of incompatibilities between the schemas
+   */
+  def compare(
+    writerSchema: Schema,
+    readerSchema: Schema,
+    writerNamedSchemas: Map[String, Schema] = Map.empty,
+    readerNamedSchemas: Map[String, Schema] = Map.empty
+  ): List[SchemaCompatibilityIssue] =
+    new SchemaComparator(writerNamedSchemas, readerNamedSchemas)
+      .compare(writerSchema, readerSchema)
+}
+
+private class SchemaComparator(
+  writerNamedSchemas: Map[String, Schema],
+  readerNamedSchemas: Map[String, Schema]
+) {
+
+  import SchemaComparator._
+
+  private val cache = new mutable.HashMap[(Schema, Schema), List[SchemaCompatibilityIssue]]
+
+  // keeps schema pairs for which comparison is ongoing in order to short circuit recursive schema comparisons
+  private val inComparison = new mutable.HashSet[(Schema, Schema)]
+
+  private def compare(writerSchema: SchemaLike, readerSchema: SchemaLike): List[SchemaCompatibilityIssue] = {
+    val normalizedWriterSchema = normalize(writerSchema, writerNamedSchemas)
+    val normalizedReaderSchema = normalize(readerSchema, readerNamedSchemas)
+    val cacheKey = (normalizedWriterSchema, normalizedReaderSchema)
+    // short-circuit recursive comparison of the same pair of schemas that is already being compared
+    if (inComparison(cacheKey)) Nil
+    else cache.getOrElseUpdate(cacheKey, {
+      inComparison.add(cacheKey)
+      try compareNormalized(normalizedWriterSchema, normalizedReaderSchema)
+      finally inComparison.remove(cacheKey)
+    })
+  }
+
+  // translate AnySchema to Schema, remove annotations and resolve references
+  @tailrec private def normalize(schema: SchemaLike, named: Map[String, Schema]): Schema = schema match {
+    case AnySchema.Anything => Schema.Empty
+    case AnySchema.Nothing => Schema.Nothing
+    case s: Schema => deannotate(s) match {
+      case s@LocalReference(name) =>
+        def noSchema: Nothing =
+          throw new NoSuchElementException(s"could not resolve schema reference ${s.$ref.get}")
+
+        normalize(named.getOrElse(name, noSchema), named)
```

Comment:
if we dereference all local schemas, then in an endpoint, if a schema is being referenced by many other schemas, a single change in that schema will cause a lot of incompatibility errors? unless we de-duplicate, e.g. by using a set, and they get combined?

Reply:
There's a cache that prevents wasting time on comparing the same pairs of schemas multiple times, but:
* cached issues will be duplicated in the final list (or tree, actually) of issues, so for better presentation we would need some additional deduplication mechanism - it could be done with some form of "references" to errors similar to schema references, or alternatively - something implemented purely in the presentation layer that displays issues to the user
* the cache is currently not reusable between toplevel schema comparisons - it would need to be extracted to some higher layer for full OpenApi comparisons

Reply:
yeah, for full OpenAPI comparisons (which is our goal), we will have to compare schemas for each endpoint. But won't simply adding the issues to a set solve the problem?

Reply:
The issues are not a flat list - they form a tree like structure that allows you to "track the path" to incompatibilities within schemas (see `SubschemaCompatibilityIssue` hierarchy)

Reply:
ah I must have missed that ... checking again ;)

Reply:
Anyway, there are several different ways we could deal with this duplication. This largely depends on how exactly do we want the incompatibilities to be presented to the user. So far, I have only implemented a simple human-readable `description` for every issue.

Reply:
Yeah, that's a good start. We'll probably want to just return the list of the issues initially.